src/lib.rs

```diff
@@ -827,23 +835,93 @@ impl SyncClient {
+                    msg
+                }
+                Either::Left((None, _)) => break,
+                Either::Right(((), _)) => break,
+            };
+
+            let mut client = wrapped_client.lock().await;
+            let response_inbox = msg.subject().to_string();
+            if let Some(mut requester_tx) = client.request_inbox_mapping.remove(&response_inbox) {
+                requester_tx.send(msg).await.unwrap_or_else(|err| {
+                    warn!("Could not write response to pending request via mapping channel. Skipping! Err: {}", err);
+                });
+            }
+        }
+
+        // At this point we are either disconnecting or we're in some
+        // strange state where the subscription's rx has been closed.
+        let mut client = wrapped_client.lock().await;
+        client.request_inbox_mapping.clear();
+        client.request_wildcard_subscription = None;
+    }
+
     async fn request(
         wrapped_client: Arc<Mutex<Self>>,
         subject: &Subject,
         payload: &[u8],
     ) -> Result<Msg> {
         let inbox_uuid = Uuid::new_v4();
-        let reply_to = format!("{}.{}", util::INBOX_PREFIX, inbox_uuid.to_simple()).parse()?;
-        let mut rx = {
+
+        let (mut rx, reply_to) = {
             let mut client = wrapped_client.lock().await;
-            let (sid, rx) = client.subscribe(&reply_to, 1).await?;
-            client.unsubscribe_with_max_msgs(sid, 1).await?;
+            let request_inbox = inbox_uuid.to_simple();
+            let reply_to: Subject = format!(
+                "{}.{}.{}",
+                util::INBOX_PREFIX,
+                client.request_base_inbox,
+                request_inbox
+            )
+            .parse()?;
+
+            // Only subscribe to the wildcard subscription when requested once!
+            if client.request_wildcard_subscription.is_none() {
+                let reply_to =
+                    format!("{}.{}.*", util::INBOX_PREFIX, client.request_base_inbox).parse()?;
+                let (sid, rx) = client.subscribe(&reply_to, 1024).await?;
+                client.request_wildcard_subscription = Some(sid);
+
+                // Spawn the task that watches the request wildcard receiver.
+                tokio::spawn(Self::request_wildcard_handler(wrapped_client.clone(), rx));
+            }
+
+            let (tx, rx) = mpsc::channel(1);
+            client
+                .request_inbox_mapping
+                .insert(reply_to.to_string(), tx);
             client
                 .publish_with_reply(subject, &reply_to, payload)
                 .await?;
-            rx
+
+            (rx, reply_to)
         };
-        Ok(rx.next().await.ok_or(Error::NoResponse)?)
+
+        // Make sure we clean up on error (don't leave a dangling request
+        // inbox mapping reference. Adding an extra mutex here seems fine
+        // since this is the error path.
+        match rx.next().await {
```

Comment:
We could use `ok_or_else` to remove the explict `match`

Reply:
I tried playing with this before but was struggling to figure out how to make the compiler happy. Doing:
```rust
        rx.next().await.ok_or_else(|| async {
            let mut client = wrapped_client.lock().await;
            client
                .request_inbox_mapping
                .remove(&reply_to)
                .or_else(|| None);
            Error::NoResponse
        })
```
Would fail with
```
    = note: expected type `std::result::Result<_, impl core::future::future::Future>`
               found type `std::result::Result<_, types::error::Error>`
    = note: the return type of a function must have a statically known size
```
I'm probably missing something obvious, but not sure what I'm missing. Pointers?

Reply:
It gets ugly but I think something like this would work

```rust
rx.next().await.ok_or_else(move || {
    tokio::spawn(async move {
        let mut client = wrapped_client.lock().await;
        client
            .request_inbox_mapping
            .remove(&reply_to.to_string())
            .or_else(|| None);
    });
    Error::NoResponse
})
```
But now that I look at it the explict spawn is proably worse than the match. Either way is fine.